app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -0,0 +1,265 @@
+        const requiresAriaLiveAnnouncement =
+          !ariaHelperText && fetchedUsers.length > 0;
+
+        if (requiresAriaLiveAnnouncement) {
+          setAriaHelperText(
+            `Mention user, ${fetchedUsers.length} results found`,
+          );
+        }
+      });
+    }
+  }, [searchTerm, fetchSuggestions, cachedSearches, ariaHelperText]);
+
+  useLayoutEffect(() => {
+    const popover = document.getElementById('mention-autocomplete-popover');
+    if (!popover) {
+      return;
+    }
+
+    const closeOnClickOutsideListener = (event) => {
+      if (!popover.contains(event.target)) {
+        // User clicked outside, reset to not searching state
+        setSearchTerm('');
+        setAriaHelperText('');
+        setUsers([]);
+      }
+    };
+
+    document.addEventListener('click', closeOnClickOutsideListener);
+
+    return () =>
+      document.removeEventListener('click', closeOnClickOutsideListener);
+  }, [searchTerm]);
+
+  useLayoutEffect(() => {
+    inputRef.current.focus();
+    inputRef.current.setSelectionRange(cursorPosition, cursorPosition - 1);
+  }, [cursorPosition]);
+
+  const handleValueChange = ({ target: { value } }) => {
+    setTextContent(value);
+    const { isUserMention, indexOfMentionStart } = getMentionWordData(
+      inputRef.current,
+    );
+
+    const { selectionStart } = inputRef.current;
+
+    if (isUserMention) {
+      // search term begins after the @ character
+      const searchTermStartPosition = indexOfMentionStart + 1;
+
+      const mentionText = value.substring(
+        searchTermStartPosition,
+        selectionStart,
+      );
+
+      const { x: cursorX, y } = getCursorXY(
+        inputRef.current,
+        indexOfMentionStart,
+      );
+      const textAreaX = inputRef.current.offsetLeft;
+
+      // On small screens always show dropdown at start of textarea
+      const dropdownX = isSmallScreen ? textAreaX : cursorX;
+
+      setDropdownPositionPoints({ x: dropdownX, y });
+      setSearchTerm(mentionText);
+      setSelectionInsertIndex(searchTermStartPosition);
+    } else if (searchTerm) {
+      // User has moved away from an in-progress @mention - clear current search
+      setSearchTerm('');
+      setAriaHelperText('');
+      setUsers([]);
+    }
+  };
+
+  const handleSelect = (username) => {
+    const textWithSelection = `${textContent.substring(
+      0,
+      selectionInsertIndex,
+    )}${username}${textContent.substring(inputRef.current.selectionStart)}`;
+
+    // Clear the current search
+    setSearchTerm('');
+    setUsers([]);
+    setAriaHelperText('');
+
+    // Update the text area value
+    setTextContent(textWithSelection);
+
+    // Update the cursor to directly after the selection
+    const newCursorPosition = selectionInsertIndex + username.length + 1;
+    setCursorPosition(newCursorPosition);
+  };
+
+  useLayoutEffect(() => {
+    if (inputRef.current) {
+      const attributes = replaceElement.attributes;
```

Comment:
Documentation (blocking): Add a comment about what we're doing here, i.e. replacing the existing textarea.

Refactor (non-blocking): Consider putting all this logic in a function

`replaceTextArea(inputRef.current, replaceElement)`

```
function replaceTextArea(node, nodeToReplace) {
...
}
```